---
title: "Create data package"
author: "Trang Tran"
date: "March 22, 2020"
output: html_notebook
---

```{r setup, include=FALSE, echo=FALSE}
library(magrittr)
library(ggplot2)
library(Biobase)

options(stringsAsFactors = FALSE)
knitr::opts_chunk$set(echo = TRUE)

file.metadata_fastq = 'fastq_metadata.tsv'
file.metadata_sra = 'PRJNA238328_metadata_cleaned.tsv'
file.metadata_pub = 'supp_file_1.tsv'
file.metadata_geo = 'samples_metadata_geo.RDS'
```

## Overview

## Download count matrices and meta data, set environment variable 'DBDIR'

```{r,eval = FALSE}
Sys.setenv('DBDIR' = '.')
```

It is assumed that files are organized into directories corresponding to the workflow, i.e. `kallisto`, `star-rsem`, `subread`.

## Experiment description - same for all data sets

```{r}
experimentData <- new("MIAME",
                      lab = "Center for Pharmacogenomics",
                      contact = "leming.shi@gmail.com",
                      title = "A rat RNA-Seq transcriptomic Bodymap across eleven organs and four developmental stages",
                      abstract = "",
                      other = list(
                          citations = c("Yu Y, Fuscoe JC, Zhao C, Guo C et al. A rat RNA-Seq transcriptomic BodyMap across 11 organs and 4 developmental stages. Nat Commun 2014;5:3230. PMID: 24510058",
                                        "Yu Y, Zhao C, Su Z, Wang C et al. Comprehensive RNA-Seq transcriptomic profiling across 11 organs, 4 ages, and 2 sexes of Fischer 344 rats. Sci Data 2014;1:140013. PMID: 25977771")
                      )
                      )
```


## STAR-RSEM workflow

### Gene-level `ExpressionSet`

#### Assay data

```{r}
gene.counts = readRDS(file.path(Sys.getenv('DBDIR'), 'star-rsem', 'matrix.gene.expected_count.RDS'))
gene.tpm = readRDS(file.path(Sys.getenv('DBDIR'), 'star-rsem', 'matrix.gene.tpm.RDS'))
gene.assayData <- new.env()
assign('exprs', gene.counts, envir = gene.assayData) # exprs is required
assign('expected_count', gene.counts, envir = gene.assayData)
assign('tpm', gene.tpm, envir = gene.assayData)
ExpressionSet(assayData = gene.assayData)
```

#### Annotations and features

```{r}
gene.featureData <- data.frame('ID' = rownames(gene.assayData$exprs))
rownames(gene.featureData) = gene.featureData$ID
gene.featureData = gene.featureData %>%
    Biobase::AnnotatedDataFrame(data = ., varMetadata = data.frame('labelDescription' = colnames(.), row.names = colnames(.)))
```

#### Phenotypic data

Biosample metadata

```{r}
METAFIELDS = data.frame(
    'name' = c('BiosampleId', 'SampleId', 'Library_Con_ng.ul', 'RNA_RIN', 'RNA_A260.A280_Ratio', 'Barcode', 'Organ'),
    'description' = c('Sample ID, in the form SAMNxxxxxxxx',
                      'Sample ID in GEO, in the form GSMxxxxxxx',
                      'Library concentration in ng/ul',
                      'RNA Integrity Number',
                      'A260/A280 ratio indicating RNA purity',
                      'The multiplexing barcode for the RNA sample',
                      'Full name of the organ from which the RNA was isolated'),
    'type' = c('character', 'character', 'numeric', 'numeric', 'numeric', 'character', 'character')
)

rownames(METAFIELDS) = METAFIELDS$name
```

```{r}
# metadata = read.table(file.path(Sys.getenv('DBDIR'),
#                                 'PRJNA238328_metadata.tsv'), sep = '\t', header = TRUE)
meta1 = read.table(file.path(Sys.getenv('DBDIR'), file.metadata_sra), sep = '\t', header = TRUE)
# Removing all the metadata fields that are conflicting for a single biosample
tmp = meta1[,setdiff(colnames(meta1),c('RunId', 'Bases', 'Bytes', 'Lane', 'AverageSpotLength'))]
meta1.bysample = meta1[!duplicated(tmp),colnames(tmp)]
str(meta1.bysample)

meta2 = read.table(file.path(Sys.getenv('DBDIR'), file.metadata_pub), sep = '\t', header = TRUE)[,c('Sample_ID', 'RNA_Sample_ID', 'ERCC_Mix', 'Barcode', 'RNA_A260.A280_Ratio', 'RNA_RIN' , 'Library_Con_ng.ul', 'Age_Week', 'Organ')]
meta2$Sample_title = paste0('SEQC_',meta2$Sample_ID)
meta2$Sample_ID <- NULL
meta3 = readRDS(file.path(Sys.getenv('DBDIR'), file.metadata_geo))
meta123 = plyr::join(meta2,meta3[,c('SampleId', 'Sample_title')]) %>%
    plyr::join(meta1.bysample, by = 'SampleId', type='left')

```

Re-processing stats

```{r}
starLog = read.table(file.path(Sys.getenv('DBDIR'), 'star-rsem',
                               'starLog.final.tsv'), sep = '\t',header=TRUE)
meta4 = data.frame(t(starLog[,-c(1,2)])) %>%
    set_colnames(paste0('starAlignment.',starLog[[1]])) %>%
    set_rownames(colnames(starLog)[-c(1,2)])
meta4$BiosampleId = rownames(meta4)
# str(meta4)
```

```{r}
libnames <- colnames(gene.assayData$exprs) %>% as.character()
allMeta <- data.frame('BiosampleId' = libnames) %>%
    plyr::join(y = meta123, by = 'BiosampleId', type = 'left')  %>%
    plyr::join(y = meta4, by = 'BiosampleId', type = 'left') %>%
    set_rownames(libnames)
colDescription = colnames(allMeta)
colDescription[grepl('starAlignment', colnames(allMeta))] = starLog[[2]]
fields_with_description = which(colnames(allMeta) %in% names(METAFIELDS))
colDescription[fields_with_description] = METAFIELDS[colDescription[fields_with_description], 'description']
for (f in colnames(allMeta)) {
    if(!is.na(METAFIELDS[f,'type'])) {
        if (METAFIELDS[f,'type'] == 'numeric') allMeta[[f]] = as.numeric(allMeta[[f]])
    }
}
phenoData <-
    Biobase::AnnotatedDataFrame(data = allMeta, varMetadata = data.frame('labelDescription' = colDescription, row.names = colnames(allMeta)))
phenoData@varMetadata
```

Making sure sample order is preserved

```{r}
all(phenoData@data$RunId == colnames(gene.counts))
```

#### Experiment description

Generated earlier, same for all data sets.

#### Assembling an `ExpressionSet`

```{r}
rbm.rnaseq.gene.star_rsem <- ExpressionSet(assayData = gene.assayData,
                             phenoData = phenoData,
                             experimentData = experimentData,
                             featureData = gene.featureData,
                             annotation = 'Ensembl Rnor_6.0.99 genome + ERCC-92'
                             )
```

### Transcript-level `ExpressionSet`

#### Assay data

```{r}
transcript.counts = readRDS(file.path(Sys.getenv('DBDIR'), 'star-rsem', 'matrix.transcripts.expected_count.RDS'))
transcript.tpm = readRDS(file.path(Sys.getenv('DBDIR'), 'star-rsem', 'matrix.transcripts.tpm.RDS'))
transcript.assayData <- new.env()
assign('exprs', transcript.counts, envir = transcript.assayData) # exprs is required
assign('expected_count', transcript.counts, envir = transcript.assayData)
assign('tpm', transcript.tpm, envir = transcript.assayData)
ExpressionSet(assayData = transcript.assayData)
```

#### Phenotypic data

Same as that of gene-level data set.

#### Annotations and features

```{r}
feature_attrs = read.table(file.path(Sys.getenv('DBDIR'), 'star-rsem', 'feature_attrs.rsem.transcripts.tsv'), sep = '\t', header=TRUE, colClasses = 'character')
transcript.featureData <- data.frame('transcript_id' = rownames(transcript.assayData$exprs)) %>%
    plyr::join(y = feature_attrs[, c('transcript_id', 'gene_id', 'length', 'effective_length')], by = 'transcript_id', type = 'left') %>%
    set_names(c('ID', 'GeneId', 'Length', 'EffectiveLength'))
rownames(transcript.featureData) = transcript.featureData$ID
transcript.featureData = transcript.featureData %>%
    Biobase::AnnotatedDataFrame(data = ., varMetadata = data.frame('labelDescription' = colnames(.), row.names = colnames(.)))
```

#### Experiment description

Generated earlier, same for all data sets.

#### Assembling an `ExpressionSet`

```{r}
rbm.rnaseq.transcript.star_rsem <- ExpressionSet(assayData = transcript.assayData,
                             phenoData = phenoData,
                             experimentData = experimentData,
                             featureData = transcript.featureData,
                             annotation = 'Ensembl Rnor_6.0.99 genome + ERCC-92'
                             )
```

## Writing data sets

```{r,eval=FALSE}
usethis::use_data(rbm.rnaseq.gene.star_rsem, internal = FALSE, overwrite = TRUE)
usethis::use_data(rbm.rnaseq.transcript.star_rsem,internal = FALSE, overwrite = TRUE)
```
